### PR TITLE
Mark QNode.backward() method as once_differentiable in the Torch interface

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Currently, QNode's are only once-differentiable, as (a) the first derivative is not-implemented in a classically end-to-end differentiable manner, and (b) higher order derivatives are not provided. This can cause some confusion when using the PyTorch interface, as PyTorch by default will just silently mark the gradient as `requires_grad=False`.

**Description of the Change:** Applies the Torch-provided decorator `@once_differentiable` to the Torch interface of the QNode backwards method, to mark it as only once-differentible.

**Benefits:** Useful error handling is now raised if attempting to differentiate quantum gradients in the Torch interface.

**Possible Drawbacks:** Once higher-order derivatives are supported, this should be removed.

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by trbromley** [comment]
Thanks @josh146 I think this would be useful! 

I tried making some code to explore the difference between what happens with and without this change and I couldn't see a difference - in both cases I got "element 0 of tensors does not require grad and does not have a grad_fn". Here is the code:

```python
import pennylane as qml
from torch.autograd import grad
import torch

dev = qml.device("default.qubit", wires=1)

@qml.qnode(dev, interface="torch")
def f(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))

x = torch.tensor(1.0, requires_grad=True)

g = grad(f(x), x, create_graph=True)[0]
g2 = grad(g, x, create_graph=True)[0]
```
Do you have any prototype code you were using?

**Comment by josh146** [comment]
@trbromley, that's weird; you should be getting the following error:

```python
RuntimeError: trying to differentiate twice a function that was marked with @once_differentiable
```

Let me go through my files and see if I can dig out my prototype.

**Comment by josh146** [comment]
Hey @trbromley, this is strange, as I can't seem to force the error message that should appear. This is the source code for the `once_differentiable` decorator:

```python
def once_differentiable(fn):

    @functools.wraps(fn)
    def wrapper(ctx, *args):
        with torch.no_grad():
            outputs = fn(ctx, *args)

        if not torch.is_grad_enabled():
            return outputs

        # If any of the inputs have requires_grad=True, we force the outputs
        # to have requires_grad=True but point to a grad_fn which throws an
        # error message during (double) back-propagation.
        # XXX: this is only an approximation of requires_grad - there's no way
        # to figure out if fn didn't use ctx.saved_tensors and as a result
        # some Tensors might require grad, even if no args do.
        # Unfortunately, this leads to unexpected error messages ("no nodes
        # require computing gradients"), but I don't have a better idea.
        # These functions would raise an error in backward anyway.
        requires_grad = any(isinstance(arg, torch.Tensor) and arg.requires_grad
                            for arg in args)
        if not requires_grad:
            return outputs

        if not isinstance(outputs, tuple):
            outputs = (outputs,)

        err_fn = torch._C._functions.DelayedError(
            b"trying to differentiate twice a function that was marked"
            b"with @once_differentiable", len(outputs))

        # Create aliases of each output that has requires_grad=True. We need
        # at least one of the inputs to err_fn to require grad so that the
        # output will have a grad_fn.
        def fake_requires_grad(var):
            if var is not None:
                var = var.detach()
                var.requires_grad = True
            return var

        return err_fn(*[fake_requires_grad(v) for v in outputs])
    return wrapper
```

and from what I can gather during debugging, `err_fn` is created... it is just never raised.

I thought that the conversions to numpy arrays in `TorchQNode.backward()` might be the issue, so I rewrote it to be a bit cleaner and avoid all numpy conversions. But still couldn't raise that issue.

File: pennylane/interfaces/torch.py [hunk 1]
```diff
@@ -144,29 +201,18 @@ def backward(ctx, grad_output):  # pragma: no cover
...
-                grad_output_np = grad_output.cpu().detach().numpy()
-            else:
-                grad_output_np = grad_output.detach().numpy()
-
-            # perform the vector-Jacobian product
-            if not grad_output_np.shape:
-                temp = grad_output_np * jacobian
-            else:
-                temp = grad_output_np.T @ jacobian
+            vjp = torch.transpose(grad_output.view(-1, 1), 0, 1) @ jacobian
```
Comment (josh146) [inline pennylane/interfaces/torch.py:206]: Note: this is actually a lot cleaner and better than what was written before, and probably also more efficient!

File: pennylane/interfaces/torch.py [hunk 2]
```diff
@@ -16,13 +16,69 @@
...
+
+    Returns:
+        Union[torch.Tensor, list, Any], array: first elements of flat arranged into the nested
+        structure of model, unused elements of flat
+    """
+    if isinstance(model, (numbers.Number, str)):
+        return flat[0], flat[1:]
+
+    if isinstance(model, torch.Tensor):
+        idx = model.size
```
Comment (trbromley) [inline pennylane/interfaces/torch.py:49]:
```suggestion
        idx = model.numel()
```
If I understand from the other version of `_unflatten` we have.

```python
x = torch.tensor(np.random.random((3, 3, 3)))
print(x.size)
print(x.size())
print(x.numel())
```

File: pennylane/interfaces/torch.py [hunk 3]
```diff
@@ -16,13 +16,69 @@
...
+    See also :func:`_flatten`.
+
+    Args:
+        flat (torch.Tensor): 1D array of items
+        model (array, Iterable, Number): model nested structure
+
+    Raises:
+        TypeError: if ``model`` contains an object of unsupported type
+
+    Returns:
```
Comment (trbromley) [inline pennylane/interfaces/torch.py:41]: Does the order of raises and returns matter? It's different for the wrapper

File: pennylane/interfaces/torch.py [hunk 4]
```diff
@@ -16,13 +16,69 @@
...
+        structure of model, unused elements of flat
+
+    Raises:
+        ValueError: if ``flat`` has more elements than ``model``
+    """
+    # pylint:disable=len-as-condition
+    res, tail = _unflatten_torch(flat, model)
+    if len(tail) != 0:
+        raise ValueError("Flattened iterable has more elements than the model.")
+    return [torch.as_tensor(i) for i in res]
```
Comment (trbromley) [inline pennylane/interfaces/torch.py:81]: I'm not sure what happens at this step?

File: pennylane/interfaces/torch.py [hunk 5]
```diff
@@ -144,29 +201,18 @@ def backward(ctx, grad_output):  # pragma: no cover
...
-            else:
-                temp = grad_output_np.T @ jacobian
+            vjp = torch.transpose(grad_output.view(-1, 1), 0, 1) @ jacobian
 
             # restore the nested structure of the input args
-            temp = [
-                np.array(i) if not isinstance(i, np.ndarray) else i
-                for i in unflatten(temp.flat, ctx.args)
-            ]
+            grad_input_list = list(unflatten_torch(vjp.flatten(), ctx.args))
```
Comment (trbromley) [inline pennylane/interfaces/torch.py:209]: Do we need this cast to a list since it seems to already happen in the function?

Comment (josh146) [reply pennylane/interfaces/torch.py:209]: Good point!
```suggestion
            grad_input_list = unflatten_torch(vjp.flatten(), ctx.args)
```

File: pennylane/interfaces/torch.py (hunk 4 above)
Comment (josh146) [reply pennylane/interfaces/torch.py:81]: The returned `res` is a list containing either scalars or lists.. This simply casts each list back to a tensor.

File: pennylane/interfaces/torch.py (hunk 3 above)
Comment (josh146) [reply pennylane/interfaces/torch.py:41]: Oops! It shouldn't (Sphinx should extract both), but this should be fixed for consistency.

**Comment by josh146** [comment]
> The above gives the usual "RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn". 🤔 Overall I guess it can't hurt to add the decorator in here though, at least it's a good reminder for developers.

This is super weird. Should we just close this PR? The other improvements to the `TorchQNode` class, while nice from a code quality point of view, were just to try and get the decorator to work, but maybe it is currently broken?

> The other changes look good, although the tests don't seem to cover the new unflatten functions, is that expected?

This is a weird issue with PyTorch custom functions; anything called from the `backwards()` method is not picked up by coverage:

```
        def backward(ctx, grad_output):  # pragma: no cover
            """Implements the backwards pass QNode vector-Jacobian product"""
            # NOTE: This method is definitely tested by the `test_torch.py` test suite,
            # however does not show up in the coverage. This is likely due to
            # subtleties in the torch.autograd.FunctionMeta metaclass, specifically
            # the way in which the backward class is created on the fly
```

**Comment by trbromley** [comment]
> This is super weird. Should we just close this PR? The other improvements to the TorchQNode class, while nice from a code quality point of view, were just to try and get the decorator to work, but maybe it is currently broken?

Thanks @josh146, I'd be in favour of closing the PR - as you said we won't need the decorator anyway when higher order derivatives are supported.

**Pull request closed without merging**